### PR TITLE
New defaults

### PR DESCRIPTION
**Pull request by ja-thomas**
So,

new defaults which seem to be better according to our simulations

- Use LCB as default infill, with lambda = 1 for numeric and lambda = 2 for mixed spaces
- New Focus search settings
- Use Imputation wrapper for default spaces
- New RF as mixed space surrogate with Jackknife SD estimation

I had to touch a lot of tests, as we switched to to LCB as default instead of just mean.

For the settings with lambda = 1 and lambda = 2 depending on the par.set, can you check if there is a nicer way to do it?

### CODE REVIEW

**Comment by ja-thomas** [comment]
Some questions:

- Do we actually want an impute wrapper for the default learner, and how exactly do we want to have it
- Should we check the infillCrit and return a GP/RF without se prediction if it is mean response?

**Comment by berndbischl** [comment]
> Do we actually want an impute wrapper for the default learner, and how exactly do we want to have it

of course. well as in our current exps. any questions here? 
(and doc this!)

> Should we check the infillCrit and return a GP/RF without se prediction if it is mean response?

for the constructed default learner? basically yes. 
but you can make this better. @jakobbossek made it possible that the infill crit has a flag whether it needs "se". so construct the learner with "se" only if the crit needs it.


> For the settings with lambda = 1 and lambda = 2 depending on the par.set, can you check if there is a nicer way to do it?

what do you mean?

**Comment by ja-thomas** [comment]
> Do we actually want an impute wrapper for the default learner, and how exactly do we want to have it
> of course. well as in our current exps. any questions here?
> (and doc this!)

No, maybe just the question how we want to impute. For numerics I set `imputeMax(2)` and for factors `imputeConstant("__miss__")`. It is documented in `makeMBOLearner` and ``mbo_default_learner`.


>Should we check the infillCrit and return a GP/RF without se prediction if it is mean response?
>for the constructed default learner? basically yes.
>but you can make this better. @jakobbossek made it possible that the infill crit has a flag whether it >needs "se". so construct the learner with "se" only if the crit needs it.

OK

>For the settings with lambda = 1 and lambda = 2 depending on the par.set, can you check if there is a nicer way to do it?
>what do you mean?

I mean this:
```
+makeMBOInfillCriterionCB = function(cb.lambda = NULL) {
...
+      if (is.null(cb.lambda))
 +        cb.lambda = ifelse(isNumeric(par.set, include.int = TRUE), 1, 2)
```
in `makeMBOInfillCriterionCB`. This doesn't look like a good solution to me but I cannot think of a better way to do it. 

We want to adapt the lambda parameter depending on the type of the search space....

**Comment by berndbischl** [comment]
> No, maybe just the question how we want to impute. For numerics I set imputeMax(2) and for factors imputeConstant("__miss__"). It is documented in makeMBOLearner and mbo_default_learner.

if it properly doced, this it totally OK for me and what i would do. i dont know something better for now.
(although we really did not benchmark this, so precisely, i cannot GUESS a better default!)

> I mean this:

the code looks perfectly clean and readable? it is basically self-commenting? what you you mean, whats bad'?

File: R/makeMBOLearner.R [hunk 1]
```diff
@@ -34,21 +42,26 @@
...
       lrn = setHyperPars(lrn, nugget.stability = 10^-8)
     else
       lrn = setHyperPars(lrn, nugget.estim = TRUE, jitter = TRUE)
   } else {
-    lrn = makeLearner("regr.randomForest", predict.type = "se", se.method = "sd")
+    lrn = makeLearner("regr.randomForest", predict.type = "se", se.method = "jackknife", keep.inbag = TRUE)
   }
+  
+  if (hasRequires(ps))
+    lrn = makeImputeWrapper(lrn, classes = list(numeric = imputeMax(2), factor = imputeConstant("__miss__")))
```
Comment (jakob-r) [inline R/makeMBOLearner.R]: If we have a numerical parameter space we want to combine imputeMax with Kriging? Is that a good idea?

I know that this case is really rare but still... for me this belongs just to Tree Based Learners.

**Comment by berndbischl** [comment]
>> If we have a numerical parameter space we want to combine imputeMax with Kriging? Is that a good idea?
>> I know that this case is really rare but still... for me this belongs just to Tree Based Learners.

thats is a good catch. the code for the "km" default is not restrictive enough. we say only "isNumeric" but I think we want isNumeric AND !hasRequires. Right?

(usually if it is purelynumeric we also have no conditional params, as conditional params are nearly without exception on categorical params)

do we want a helper for that in PH? then give me a good name. 
@ja-thomas

**Comment by ja-thomas** [comment]
Good point, thanks.

Do we actually need a helper function for that? It is basically just `isNumeric(x) & !hasRequires(x)`.

**Comment by jakob-r** [comment]
I am happy with hot it is now.

**Comment by berndbischl** [comment]
yes its very "hot", lets merge it

**Pull request merged**